File: krueger/webapp/src/io/eleven19/krueger/webapp/monaco/ElmLanguage.scala
package io.eleven19.krueger.webapp.monaco

import scala.scalajs.js

import io.eleven19.krueger.lexer.ElmTokenKind
import io.eleven19.krueger.lexer.ElmTokenizer

/** Monaco language descriptor for the Elm dialect Krueger parses.
  *
  * Highlighting is sourced from the shared core tokenizer so parser-facing and UI-facing token categories stay aligned.
  *
  * All Monaco touch-points live behind [[register]], keeping the rest of this object callable from Node (where
  * `monaco-editor` is not resolvable) so tests can assert on the pure token adapter.
  */
object ElmLanguage:

    val id: String = "elm"

    final case class TokenSpan(startIndex: Int, scope: String) derives CanEqual

    def scopeFor(kind: ElmTokenKind): String = kind match
        case ElmTokenKind.Keyword         => "keyword"
        case ElmTokenKind.LowerIdentifier => "identifier"
        case ElmTokenKind.UpperIdentifier => "type.identifier"
        case ElmTokenKind.Operator        => "operator"
        case ElmTokenKind.Number          => "number"
        case ElmTokenKind.StringLiteral   => "string"
        case ElmTokenKind.CharLiteral     => "string"
        case ElmTokenKind.Comment         => "comment"
        case ElmTokenKind.Whitespace      => "white"
        case ElmTokenKind.Newline         => "white"
        case ElmTokenKind.Punctuation     => "delimiter"
        case ElmTokenKind.Unknown         => "invalid"

    def tokenSpans(line: String): Vector[TokenSpan] =
        val result = ElmTokenizer.run(line)
        result.value.fold(
            _ => Vector(TokenSpan(0, "invalid")),
            tokens => tokens.filterNot(_.kind == ElmTokenKind.Newline).map(t => TokenSpan(t.start, scopeFor(t.kind)))
        )

    lazy val tokensProvider: MonacoFacade.TokensProvider = new MonacoFacade.TokensProvider:
        private lazy val state: MonacoFacade.IState =
            js.Dynamic.literal(
                clone = () => state,
                equals = (_: MonacoFacade.IState) => true
            ).asInstanceOf[MonacoFacade.IState]

        def getInitialState(): MonacoFacade.IState = state

        def tokenize(line: String, previousState: MonacoFacade.IState): MonacoFacade.ILineTokens =
            js.Dynamic.literal(
                endState = state,
                tokens = js.Array(tokenSpans(line).map(span =>
                    js.Dynamic.literal(startIndex = span.startIndex, scopes = span.scope).asInstanceOf[MonacoFacade.IToken]
                )*)
            ).asInstanceOf[MonacoFacade.ILineTokens]

    /** Idempotently register Elm with Monaco. Safe to call twice — the second call is a no-op because Monaco's
      * `getLanguages` will already list the id.
      *
      * @note
      *   Must only be called in a browser context. Unit tests must not invoke this method; they assert on the pure data
      *   above.
      */
    def register(): Unit =
        val monaco       = MonacoFacade.Monaco
        val alreadyThere = monaco.languages.getLanguages().exists(_.id == id)
        if !alreadyThere then
            val ext = js.Dynamic.literal(id = id).asInstanceOf[MonacoFacade.LanguageExtensionPoint]
            monaco.languages.register(ext)
            monaco.languages.setTokensProvider(id, tokensProvider)

File: sites/try-wasm/src/lib/elm-language.ts
type MonacoModule = typeof import('monaco-editor/esm/vs/editor/editor.api');

import type { CompilerEnvelope } from './krueger';

export const elmLanguageId = 'elm';

export type ElmTokenKind =
  | 'Keyword'
  | 'LowerIdentifier'
  | 'UpperIdentifier'
  | 'Operator'
  | 'Number'
  | 'StringLiteral'
  | 'CharLiteral'
  | 'Comment'
  | 'Whitespace'
  | 'Newline'
  | 'Punctuation'
  | 'Unknown';

export type ElmToken = {
  kind: ElmTokenKind;
  lexeme: string;
  start: number;
  end: number;
};

export type ElmTokenizer = (line: string) => CompilerEnvelope<ElmToken[]>;

export type MonacoTokenSpan = {
  startIndex: number;
  scopes: string;
};

export function scopeFor(kind: ElmTokenKind): string {
  switch (kind) {
    case 'Keyword':
      return 'keyword';
    case 'LowerIdentifier':
      return 'identifier';
    case 'UpperIdentifier':
      return 'type.identifier';
    case 'Operator':
      return 'operator';
    case 'Number':
      return 'number';
    case 'StringLiteral':
    case 'CharLiteral':
      return 'string';
    case 'Comment':
      return 'comment';
    case 'Whitespace':
    case 'Newline':
      return 'white';
    case 'Punctuation':
      return 'delimiter';
    case 'Unknown':
      return 'invalid';
  }
}

export function tokenSpans(line: string, tokenize: ElmTokenizer = globalTokenizer): MonacoTokenSpan[] {
  const result = tokenize(line);
  if (!result.ok || result.value === null) return [{ startIndex: 0, scopes: 'invalid' }];

  return result.value
    .filter((token) => token.kind !== 'Newline')
    .map((token) => ({ startIndex: token.start, scopes: scopeFor(token.kind) }));
}

export function createElmTokensProvider(tokenize: ElmTokenizer = globalTokenizer): MonacoModule['languages']['TokensProvider'] {
  const state = {
    clone: () => state,
    equals: () => true
  };

  return {
    getInitialState: () => state,
    tokenize: (line) => ({
      endState: state,
      tokens: tokenSpans(line, tokenize)
    })
  };
}

export function registerElmLanguage(monaco: MonacoModule, tokenize: ElmTokenizer = globalTokenizer): void {
  const alreadyRegistered = monaco.languages
    .getLanguages()
    .some((language) => language.id === elmLanguageId);

  if (alreadyRegistered) return;

  monaco.languages.register({ id: elmLanguageId });
  monaco.languages.setTokensProvider(elmLanguageId, createElmTokensProvider(tokenize));
}

function globalTokenizer(line: string): CompilerEnvelope<ElmToken[]> {
  const facade = (globalThis as { Krueger?: { tokenize?: (source: string) => unknown } }).Krueger;
  if (typeof facade?.tokenize !== 'function') {
    return {
      ok: false,
      value: null,
      logs: [],
      errors: [{ phase: 'tokenize', message: 'Krueger tokenizer facade is not loaded' }]
    };
  }

  return normalizeTokenEnvelope(facade.tokenize(line));
}

function normalizeTokenEnvelope(raw: unknown): CompilerEnvelope<ElmToken[]> {
  const record = asRecord(raw);
  const ok = Boolean(record.ok);
  const value = ok && Array.isArray(record.value) ? record.value.map(normalizeToken) : null;

  return {
    ok,
    value,
    logs: Array.isArray(record.logs) ? record.logs.map(String) : [],
    errors: []
  };
}

function normalizeToken(raw: unknown): ElmToken {
  const record = asRecord(raw);
  return {
    kind: normalizeKind(record.kind),
    lexeme: String(record.lexeme ?? ''),
    start: Number(record.start ?? 0),
    end: Number(record.end ?? 0)
  };
}

function normalizeKind(value: unknown): ElmTokenKind {
  const rendered = String(value);
  if (isElmTokenKind(rendered)) return rendered;
  return 'Unknown';
}

function isElmTokenKind(value: string): value is ElmTokenKind {
  return (
    value === 'Keyword' ||
    value === 'LowerIdentifier' ||
    value === 'UpperIdentifier' ||
    value === 'Operator' ||
    value === 'Number' ||
    value === 'StringLiteral' ||
    value === 'CharLiteral' ||
    value === 'Comment' ||
    value === 'Whitespace' ||
    value === 'Newline' ||
    value === 'Punctuation' ||
    value === 'Unknown'
  );
}

function asRecord(value: unknown): Record<string, unknown> {
  return value !== null && typeof value === 'object' ? (value as Record<string, unknown>) : {};
}
